### PR TITLE
[2.x] feat(revisions): Allow draft revisions on top of published content

### PR DESCRIPTION
**Pull request by pboivin**
## Description

This PR introduces a new `$enableDraftRevisions` controller option. The option allows revisions to be added on top of published items, without actually modifying the content of the record in the database.

The feature is disabled by default, so the default behavior of module revisions is not changed.

Here's a quick demonstration of the workflow:

https://user-images.githubusercontent.com/7805679/176737937-48650087-f621-4864-85de-7fc49b88fe98.mp4

## Related Issues

Based on/includes https://github.com/area17/twill/pull/1719

Here are the commits pertaining to this feature : 
https://github.com/area17/twill/pull/1725/files/b86ae5363839537e4de2ecd36c01ac3531e05871..d8b4f559c5ef20019b17fdf4f2dbf2c1a85d0c54

### CODE REVIEW

File: src/Http/Controllers/Admin/Concerns/FormSubmitOptions.php [hunk 1]
```diff
@@ -0,0 +1,254 @@
+<?php
+
+namespace A17\Twill\Http\Controllers\Admin\Concerns;
+
+use Illuminate\Database\Eloquent\Model;
+
+trait FormSubmitOptions
```
Comment (haringsrob) [inline src/Http/Controllers/Admin/Concerns/FormSubmitOptions.php:7]: Maybe we can rename this to "HasDefaultFormSubmitOptions"?

File: src/Http/Controllers/Admin/ModuleController.php [hunk 2]
```diff
@@ -250,6 +252,13 @@ abstract class ModuleController extends Controller
      */
     protected $fieldsPermissions = [];
 
+    /**
+     * Determines if draft revisions can be added on top of published content.
+     *
+     * @var bool
+     */
+    protected $enableDraftRevisions = false;
```
Comment (haringsrob) [inline src/Http/Controllers/Admin/ModuleController.php:260]: Can we add a todo to make a setter for this in 3.x?

File: src/Models/Behaviors/HasRevisions.php [hunk 3]
```diff
@@ -36,14 +36,22 @@ public function scopeMine($query)
...
+            ->map(function ($revision, $index) use (&$currentRevision) {
+                if (!$currentRevision && !$revision->isDraft()) {
+                    $currentRevision = $revision;
+                }
+
+                return [
+                    'id' => $revision->id,
+                    'author' => $revision->user->name ?? 'Unknown',
+                    'datetime' => $revision->created_at->toIso8601String(),
+                    'label' => $currentRevision === $revision ? twillTrans('twill::lang.publisher.current') : '',
```
Comment (haringsrob) [inline src/Models/Behaviors/HasRevisions.php:51]: I wonder how "safe" this comparison is: `$currentRevision === $revision`

Perhaps we can compare the id's?

File: src/Http/Controllers/Admin/Concerns/FormSubmitOptions.php (hunk 1 above)
Comment (pboivin) [reply src/Http/Controllers/Admin/Concerns/FormSubmitOptions.php:7]: This is a bit nitpick-y, but here are my reasons for keeping the curent name:

1. In Twill, I believe the `HasX` naming is typically used for model traits (aka. behaviors). I would prefer to have a different convention for controller traits, hence the `Concerns\X` structure.

2. `HasDefaultFormSubmitOptions` makes sense with regards to [#1719](https://github.com/area17/twill/pull/1719) but I think this PR goes a bit beyond the "defaults" by introducing the variations for the context of draft revisions.

File: src/Models/Behaviors/HasRevisions.php (hunk 3 above)
Comment (pboivin) [reply src/Models/Behaviors/HasRevisions.php:51]: I'm not sure I follow. Can you expand on this a bit more?

Here, we start from a fixed list of object references ($this->revisions), and try to identify which one should have the `Current` label. In my view, comparing object references is not a problem.

FWIW, there a 2 tests covering this method: [(1)](https://github.com/area17/twill/pull/1725/files/d8b4f559c5ef20019b17fdf4f2dbf2c1a85d0c54#diff-695f6fd51bb71c5db4574813ec07caefc86109a094bcd60b28b482c17776c81dR119) [(2)](https://github.com/area17/twill/pull/1725/files/d8b4f559c5ef20019b17fdf4f2dbf2c1a85d0c54#diff-695f6fd51bb71c5db4574813ec07caefc86109a094bcd60b28b482c17776c81dR189)

File: src/Http/Controllers/Admin/ModuleController.php (hunk 2 above)
Comment (pboivin) [reply src/Http/Controllers/Admin/ModuleController.php:260]: Yes, feel free to add any anotations you would find helpful.

**Pull request merged**